File: Day24_R_Programming/titanic_eda.R
# ==========================================
# DAY 24 - TITANIC EDA PROJECT
# Exploratory Data Analysis in R
# ==========================================


# ------------------------------------------
# 1. LOAD PACKAGES
# ------------------------------------------

library(dplyr)
library(tidyr)
library(ggplot2)


# ------------------------------------------
# 2. LOAD DATASET
# ------------------------------------------

titanic <- read.csv("titanic.csv")


# ------------------------------------------
# 3. INITIAL DATA EXPLORATION
# ------------------------------------------

# First 6 rows
head(titanic)

# Number of rows and columns
dim(titanic)

# Column names
colnames(titanic)

# Structure of dataset
str(titanic)

# Summary statistics
summary(titanic)


# ------------------------------------------
# 4. CHECK MISSING VALUES
# ------------------------------------------

colSums(is.na(titanic))


# ------------------------------------------
# 5. DATA CLEANING
# ------------------------------------------

# Fill missing Age values with mean Age
# Fill missing Fare using median Fare for the
# corresponding passenger class

titanic_clean <- titanic %>%
  mutate(
    Age = ifelse(
      is.na(Age),
      mean(Age, na.rm = TRUE),
      Age
    ),
    Fare = ifelse(
      is.na(Fare),
      median(Fare[Pclass == 3], na.rm = TRUE),
      Fare
    )
  )


# Check missing values after cleaning

colSums(is.na(titanic_clean))


# ------------------------------------------
# 6. OVERALL SURVIVAL ANALYSIS
# ------------------------------------------

# Number of survivors and non-survivors

titanic_clean %>%
  count(Survived)


# Overall survival rate

titanic_clean %>%
  summarize(
    survival_rate = mean(Survived) * 100
  )


# ------------------------------------------
# 7. SURVIVAL BY GENDER
# ------------------------------------------

titanic_clean %>%
  group_by(Sex) %>%
  summarize(
    passengers = n(),
    survivors = sum(Survived),
    survival_rate = mean(Survived) * 100,
    .groups = "drop"
  )


# ------------------------------------------
# 8. SURVIVAL BY PASSENGER CLASS
# ------------------------------------------

titanic_clean %>%
  group_by(Pclass) %>%
  summarize(
    passengers = n(),
    survivors = sum(Survived),
    survival_rate = mean(Survived) * 100,
    .groups = "drop"
  )


# ------------------------------------------
# 9. AGE ANALYSIS
# ------------------------------------------

# Overall average age

titanic_clean %>%
  summarize(
    average_age = mean(Age)
  )


# Average age by gender

titanic_clean %>%
  group_by(Sex) %>%
  summarize(
    average_age = mean(Age),
    .groups = "drop"
  )


# Average age by passenger class

titanic_clean %>%
  group_by(Pclass) %>%
  summarize(
    average_age = mean(Age),
    .groups = "drop"
  )


# ------------------------------------------
# 10. SURVIVAL BY CLASS AND GENDER
# ------------------------------------------

titanic_clean %>%
  group_by(Pclass, Sex) %>%
  summarize(
    passengers = n(),
    survivors = sum(Survived),
    survival_rate = mean(Survived) * 100,
    .groups = "drop"
  )


# ------------------------------------------
# 11. VISUALIZATION - SURVIVAL BY GENDER
# ------------------------------------------

ggplot(
  titanic_clean,
  aes(x = Sex, fill = factor(Survived))
) +
  geom_bar(position = "dodge") +
  labs(
    title = "Survival by Gender",
    x = "Gender",
    y = "Number of Passengers",
    fill = "Survived"
  )


# ------------------------------------------
# 12. VISUALIZATION - AGE DISTRIBUTION
# ------------------------------------------

ggplot(
  titanic_clean,
  aes(x = Age)
) +
  geom_histogram(bins = 20) +
  labs(
    title = "Age Distribution of Titanic Passengers",
    x = "Age",
    y = "Number of Passengers"
  )


# ------------------------------------------
# 13. VISUALIZATION - SURVIVAL BY CLASS
# ------------------------------------------

ggplot(
  titanic_clean,
  aes(
    x = factor(Pclass),
    fill = factor(Survived)
  )
) +
  geom_bar(position = "dodge") +
  labs(
    title = "Survival by Passenger Class",
    x = "Passenger Class",
    y = "Number of Passengers",
    fill = "Survived"
  )


# ------------------------------------------
# 14. VISUALIZATION - AGE BY SURVIVAL
# ------------------------------------------

ggplot(
  titanic_clean,
  aes(
    x = Age,
    fill = factor(Survived)
  )
) +
  geom_histogram(
    bins = 20,
    position = "identity",
    alpha = 0.6
  ) +
  labs(
    title = "Age Distribution by Survival",
    x = "Age",
    y = "Number of Passengers",
    fill = "Survived"
  )



# ------------------------------------------
# Key Insights:

- Most passengers were between approximately 20 and 40 years old.
- Both survivors and non-survivors were present across most age groups.
- Age alone did not show a clear survival pattern.
- Non-survivors were generally more numerous than survivors across many age groups.
# ------------------------------------------


# ------------------------------------------
# 15. END OF PROJECT
# ------------------------------------------

# Day 24 Titanic EDA completed

File: Project1_EDA_Titanic/titanic_eda.py
# ==========================================
# PROJECT 1 - TITANIC EDA PROJECT
# Exploratory Data Analysis in Python
# ==========================================


# ------------------------------------------
# 1. LOAD PACKAGES
# ------------------------------------------

import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns


# ------------------------------------------
# 2. LOAD DATASET
# ------------------------------------------

titanic = pd.read_csv("titanic.csv")


# ------------------------------------------
# 3. INITIAL DATA EXPLORATION
# ------------------------------------------

print(titanic.head())
print(titanic.shape)
print(titanic.columns)
print(titanic.info())
print(titanic.describe())


# ------------------------------------------
# 4. CHECK MISSING VALUES
# ------------------------------------------

print(titanic.isna().sum())


# ------------------------------------------
# 5. DATA CLEANING
# ------------------------------------------

age_mean = titanic["Age"].mean()
titanic_clean = titanic.copy()
titanic_clean["Age"] = titanic_clean["Age"].fillna(age_mean)

fare_median_class3 = titanic_clean.loc[
    titanic_clean["Pclass"] == 3,
    "Fare"
].median()

titanic_clean["Fare"] = titanic_clean["Fare"].fillna(
    fare_median_class3
)

print(titanic_clean.isna().sum())


# ------------------------------------------
# 6. OVERALL SURVIVAL ANALYSIS
# ------------------------------------------

print(titanic_clean["Survived"].value_counts())

survival_rate = titanic_clean["Survived"].mean() * 100
print("Overall Survival Rate:", survival_rate)


# ------------------------------------------
# 7. SURVIVAL BY GENDER
# ------------------------------------------

gender_survival = (
    titanic_clean
    .groupby("Sex")
    .agg(
        passengers=("Survived", "size"),
        survivors=("Survived", "sum"),
        survival_rate=("Survived", "mean")
    )
    .reset_index()
)
gender_survival["survival_rate"] = gender_survival["survival_rate"] * 100
print(gender_survival)


# ------------------------------------------
# 8. SURVIVAL BY PASSENGER CLASS
# ------------------------------------------

class_survival = (
    titanic_clean
    .groupby("Pclass")
    .agg(
        passengers=("Survived", "size"),
        survivors=("Survived", "sum"),
        survival_rate=("Survived", "mean")
    )
    .reset_index()
)
class_survival["survival_rate"] = class_survival["survival_rate"] * 100
print(class_survival)


# ------------------------------------------
# 9. AGE ANALYSIS
# ------------------------------------------

average_age = titanic_clean["Age"].mean()
print("Overall Average Age:", average_age)

average_age_gender = (
    titanic_clean.groupby("Sex")["Age"].mean().reset_index()
)
print(average_age_gender)

average_age_class = (
    titanic_clean.groupby("Pclass")["Age"].mean().reset_index()
)
print(average_age_class)


# ------------------------------------------
# 10. SURVIVAL BY CLASS AND GENDER
# ------------------------------------------

class_gender_survival = (
    titanic_clean
    .groupby(["Pclass", "Sex"])
    .agg(
        passengers=("Survived", "size"),
        survivors=("Survived", "sum"),
        survival_rate=("Survived", "mean")
    )
    .reset_index()
)
class_gender_survival["survival_rate"] = class_gender_survival["survival_rate"] * 100
print(class_gender_survival)


# ------------------------------------------
# 11. VISUALIZATION - SURVIVAL BY GENDER
# ------------------------------------------

sns.countplot(data=titanic_clean, x="Sex", hue="Survived")
plt.title("Survival by Gender")
plt.xlabel("Gender")
plt.ylabel("Number of Passengers")
plt.legend(title="Survived")
plt.show()


# ------------------------------------------
# 12. VISUALIZATION - AGE DISTRIBUTION
# ------------------------------------------

plt.hist(titanic_clean["Age"], bins=20)
plt.title("Age Distribution of Titanic Passengers")
plt.xlabel("Age")
plt.ylabel("Number of Passengers")
plt.show()


# ------------------------------------------
# 13. VISUALIZATION - SURVIVAL BY CLASS
# ------------------------------------------

sns.countplot(data=titanic_clean, x="Pclass", hue="Survived")
plt.title("Survival by Passenger Class")
plt.xlabel("Passenger Class")
plt.ylabel("Number of Passengers")
plt.legend(title="Survived")
plt.show()


# ------------------------------------------
# 14. VISUALIZATION - AGE BY SURVIVAL
# ------------------------------------------

sns.histplot(
    data=titanic_clean,
    x="Age",
    hue="Survived",
    bins=20,
    multiple="layer",
    alpha=0.6
)
plt.title("Age Distribution by Survival")
plt.xlabel("Age")
plt.ylabel("Number of Passengers")
plt.show()


# ------------------------------------------
# Key Insights:
#
# - Females had a significantly higher survival rate than males
# - 1st class passengers had higher survival rate than 3rd class
# - Most passengers were between approximately 20 and 40 years old
# - Age alone did not show a clear survival pattern
# ------------------------------------------


# ------------------------------------------
# 15. END OF PROJECT
# ------------------------------------------

print("Project 1 Titanic EDA completed!")
